### ISSUE TITLE
Simpler usage for PreprocessorPyTorch

### ISSUE THREAD

**Issue by davidslater** [Trusted-AI/adversarial-robustness-toolbox#708]
**Is your feature request related to a problem? Please describe.**
To create a very pytorch-specific preprocessor, you have to write a lot of code (some of which includes numpy and some of which includes pytorch). Here is an example of truncating a video:
```
from art.defences.preprocessor.preprocessor import PreprocessorPyTorch                   
                                                                                         
class VideoTruncate(PreprocessorPyTorch):                                                
    """                                                                                  
    Simple PyTorch preprocessing (truncate frames from video)                            
    """                                                                                  
                                                                                         
    def __init__(self, max_frames=64, device_type="gpu"):                                
        import torch                                                                     
        super().__int__()                                                                
        max_frames = int(max_frames)                                                     
        if max_frames < 0:                                                               
            raise ValueError(f"max_frames {max_frames} cannot be negative")              
        self.max_frames = max_frames                                                     
        if device_type == "cpu" or not torch.cuda.is_available():                        
            self._device = torch.device("cpu")                                           
        else:                                                                            
            cuda_idx = torch.cuda.current_device()                                       
            self._device = torch.device("cuda:{}".format(cuda_idx))                      
                                                                                         
    def forward(self, x, y=None):                                                        
        """                                                                              
        Assuming video shape (batch_size, frames, height, width, channels)               
        """                                                                              
        return x[:, :self.max_frames], y                                                 
                                                                                         
    def estimate_forward(self, x, y=None):                                               
        return self.forward(x, y=y)

    def __call__(self, x, y=None):                                                       
        x = torch.tensor(x, device=self._device)                                         
        if y is not None:                                                                
            y = torch.tensor(y, device=self._device)                                     
        with torch.no_grad():                                                            
            x, y = self.forward(x, y)                                                    
                                                                                         
        x = x.cpu().numpy()                                                              
        if y is not None:                                                                
            y = y.cpu().numpy()                                                          
                                                                                         
        return x, y                                                                      
                                                                                         
    def estimate_gradient(self, x, gradients):                                           
        x = torch.tensor(x, device=self._device, requires_grad=True)                     
        gradients = torch.tensor(gradients, device=self._device)                         
        x_orig = x                                                                       
        x = self.estimate_forward(x)                                                     
        x.backward(gradients)                                                            
        gradients = x_orig.grad.detach().cpu().numpy()                                   
        return gradients                                                                 
                                                                                         
    def apply_fit(self):                                                                 
        return True                                                                      
                                                                                         
    def apply_predict(self):                                                             
        return True                                                                      
                                                                                         
    def fit(self):                                                                       
        pass
```

**Describe the solution you'd like**
I would like to be able to write the following code instead (only the forward pass should be required), which would be very intuitive for those who know pytorch:
```
class VideoTruncate(PreprocessorPyTorch):                                                
    """                                                                                  
    Simple PyTorch preprocessing (truncate frames from video)                            
    """                                                                                  
                                                                                         
    def __init__(self, max_frames=64):                                                   
        super().__int__()                                                                
        max_frames = int(max_frames)                                                     
        if max_frames < 0:                                                               
            raise ValueError(f"max_frames {max_frames} cannot be negative")              
        self.max_frames = max_frames                                                     
                                                                                         
    def forward(self, x, y=None):                                                        
        """                                                                              
        Assuming video shape (batch_size, frames, height, width, channels)               
        """                                                                              
        return x[:, :self.max_frames], y
```

This requires modifications to `PreprocessorPyTorch`. 

Ideally, this would also change `PyTorchEstimator ` to work consistently when preprocessing defenses has a list of size 1. Similar modifications would be helpful on the TensorFlow side as well.

**Describe alternatives you've considered**
None. Currently, things work, but are rather clunky and require more deeply understanding the subclasses. With this approach, you could reduce the user knowledge and effort required, and ensure greater consistency for pytorch and numpy interfaces. 

**Additional context**
If this is of interest, I am happy to contribute. Just let me know, and point me to the milestone you would want it in.

**Comment by beat-buesser** [comment]
Hi @davidslater Thank you very much for your suggesting this feature and for your interest to contribute it. I like that your suggestions looks much cleaner. I agree for the framework-specific preprocessing we don't really need `__call__`, `estimate_forward` and `estimate_gradient` anymore, I think they were added for backward compatibility in earlier versions.

We'll need to consider `apply_fit`, `apply_predict`, `fit` as they help distinguishing different kinds of preprocessing, but maybe it's possible to define defaults for them covering the most common cases in the base class so that they only have to be implemented in special cases.

We have currently PR #629 which combines `preprocessing_defences` and `preprocessing` into a single framework-specific application of preprocessing which enables custom `preprocessing` (normalisation) and gradient calculation through model and preprocessing in one backward pass. If you could take into account the changes coming with #629 and make a pull request maybe a week before the release for 1.5.0 on Nov. 30, or earlier, we should be able to include it in the next release.

What do you think?

**Comment by davidslater** [comment]
That sounds good. Can you assign this to me?

**Comment by beat-buesser** [comment]
Great, thank you! I assigned the issue and added it to the ART 1.5.0 project and milestone.
